File: src/foam/dao/SlowDAO.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.dao',
  name: 'SlowDAO',
  extends: 'foam.dao.ProxyDAO',
  properties: [
    {
      class: 'Int',
      name: 'delayMs',
    },
  ],
  methods: [
    {
      name: 'select_',
      swiftCode: `
usleep(UInt32(delayMs * 1000))
return try super.select_(x, sink, skip, limit, order, predicate)
      `,
      javaCode: `
try {
  Thread.sleep(getDelayMs());
} catch(InterruptedException e) {
}

return super.select_(x, sink, skip, limit, order, predicate);
      `,
    },
    {
      name: 'put_',
      swiftCode: `
usleep(UInt32(delayMs * 1000))
return try super.put_(x, obj)
      `,
      javaCode: `
try {
  Thread.sleep(getDelayMs());
} catch(InterruptedException e) {
}

return super.put_(x, obj);
      `,
    },
    {
      name: 'remove_',
      swiftCode: `
usleep(UInt32(delayMs * 1000))
return try super.remove_(x, obj)
      `,
      javaCode: `
try {
  Thread.sleep(getDelayMs());
} catch(InterruptedException e) {
}

return super.remove_(x, obj);
      `,
    },
    {
      name: 'removeAll_',
      swiftCode: `
usleep(UInt32(delayMs * 1000))
return try super.removeAll_(x)
      `,
      javaCode: `
try {
  Thread.sleep(getDelayMs());
} catch(InterruptedException e) {
}

super.removeAll_(x, skip, limit, order, predicate);
      `,
    },
    {
      name: 'find_',
      swiftCode: `
usleep(UInt32(delayMs * 1000))
return try super.find_(x, id)
      `,
    },
  ]
});

File: src/foam/dao/SlowDAO.java
package foam.dao;

public class SlowDAO
  extends ProxyDAO
{
  private int delayms_ = 500;

  public SlowDAO(DAO delegate) {
    super();
    setDelegate(delegate);
  }

  public SlowDAO(int delayms, DAO delegate) {
    this(delegate);
    delayms_ = delayms;
  }

  @Override
  public foam.core.FObject put_(foam.core.X x, foam.core.FObject obj) {
    try {
      Thread.sleep(delayms_);
    } catch(InterruptedException e) {
    }

    return super.put_(x, obj);
  }

  @Override
  public foam.core.FObject remove_(foam.core.X x, foam.core.FObject obj) {
    try {
      Thread.sleep(delayms_);
    } catch(InterruptedException e) {
    }

    return super.remove_(x, obj);
  }

  @Override
  public foam.dao.Sink select_(foam.core.X x, foam.dao.Sink sink, long skip, long limit, foam.mlang.order.Comparator order, foam.mlang.predicate.Predicate predicate) {
    try {
      Thread.sleep(delayms_);
    } catch(InterruptedException e) {
    }

    return super.select_(x, sink, skip, limit, order, predicate);
  }

  @Override
  public void removeAll_(foam.core.X x, long skip, long limit, foam.mlang.order.Comparator order, foam.mlang.predicate.Predicate predicate) {
    try {
      Thread.sleep(delayms_);
    } catch(InterruptedException e) {
    }

    super.removeAll_(x, skip, limit, order, predicate);
  }
}
